lib/sms/sms_at.c

```diff
@@ -46,6 +46,52 @@ LOG_MODULE_DECLARE(sms, CONFIG_SMS_LOG_LEVEL);
 /** @brief Length of AT notification for incoming SMS status report. */
 #define AT_SMS_STATUS_REPORT_LEN (sizeof(AT_SMS_STATUS_REPORT) - 1)
 
+/**
+ * @brief SMS-STATUS-REPORT PDU in 3GPP TS 23.040 chapter 9.2.2.3
+ */
+#define SMS_STATUS_REPORT_HEADER_OFFSET   0x00
+#define SMS_STATUS_REPORT_TP_MR_OFFSET    0x01
+#define SMS_STATUS_REPORT_TP_RA_OFFSET    0x02
+#define SMS_STATUS_REPORT_TP_SCTS_LEN     7
+#define SMS_STATUS_REPORT_TP_DT_LEN       7
+#define SMS_STATUS_REPORT_TP_ST_LEN       1
+
+/**
+ * @brief TP-Message-Type-Indicator in 3GPP TS 23.040 chapter 9.2.3.1
+ */
+#define TP_MTI_SMS_DELIVERY         0x00
+#define TP_MTI_SMS_DELIVERY_REPORT  0x00
+#define TP_MTI_SMS_STATUS_REPORT    0x02
+#define TP_MTI_SMS_COMMAND          0x02
+#define TP_MTI_SMS_SUBMIT           0x01
+#define TP_MTI_SMS_SUBMIT_REPORT    0x01
+#define TP_MTI_RESERVED             0x03
+
+/**
+ * @brief TP-Status in 3GPP TS 23.040 chapter 9.2.3.15
+ */
+#define TP_STATUS_RECEIVED          0x00
+#define TP_STATUS_FORWARDED         0x01
+#define TP_STATUS_REPLACED          0x02
+#define TP_STATUS_SC_SPECIFIC(status) (status & 0x10)
+
+/**
+ * @brief TP-SRQ in 3GPP TS 23.040 chapter 9.2.3.26
+ */
+#define TP_SRQ_SMS_SUBMIT           0x00
+#define TP_SRQ_SMS_COMMAND          0x01
+
+/**
+ * @brief First byte of SMS-STATUS-REPORT PDU in 3GPP TS 23.040 chapter 9.2.2.3
+ */
+struct pdu_status_report_header {
```

Comment:
I've only had a quick look at this from higher level.
I think we should do SMS-STATUS-REPORT decoding in the same way as SMS-DELIVER message, i.e., decode various fields and have a separate module/file for that like sms_deliver.c. And `union sms_header` should have an additional field with something like `struct sms_status_report_header status_report`

Reply:
And whatever additions you do for SMS library, please create tests under tests/lib/sms. Currently SMS library is tested with ~90% branch coverage so let's try to keep that.

Reply:
Understood, will create new code file. Some functions in sms_deliver.c would be shared in some way.

Reply:
@trantanen added "sms_status_report.c" to parse SMS_STATUS_REPORT, with below limitation
`       /* Only parse SMS-STATUS-REPORT for SMS-SUBMIT */
	if (pdata->header.mti != TP_MTI_SMS_STATUS_REPORT ||
	    pdata->header.srq != TP_SRQ_SMS_SUBMIT) {
		return -EINVAL;
	}`
Only mandatory fields of the PDU header are parsed.
Please review.